### PR TITLE
feat(modelexpress): add S3 delta weight update backend

### PR DESCRIPTION
**Pull request by nv-hwoo**
## Summary

  This PR adds ModelExpress as a VIME weight-update backend, with initial support for S3-based canonical
  delta weight synchronization.

  Instead of publishing a complete checkpoint after every training update, the trainer captures an initial
  Hugging Face baseline and publishes subsequent policy updates as versioned XOR deltas through
  ModelExpress.

```mermaid
  flowchart LR
      T[Megatron trainer] -->|canonical HF weights| MX[ModelExpress trainer client]
      MX -->|XOR delta artifacts| S3[S3-compatible storage]
      MX -->|version lifecycle| C[ModelExpress catalog]
      C --> V[vLLM ModelExpress backend]
      S3 --> V
```
## What changed
- Adds modelexpress as a valid --update-weight-transport.
- Lazily imports the ModelExpress implementation only when that transport is selected.
- Configures vLLM to use its modelexpress weight-transfer backend.
- Converts Megatron weights into canonical Hugging Face tensor chunks.
- Distributes chunk ownership across trainer ranks so each chunk is published once.
- Automatically creates the catalog-only, READY v0 baseline before initializing rollout engines.
- Captures the initial seed checkpoint as the trainer’s delta baseline.
- Publishes later versions as:
    - S3 XOR deltas by default.
    - Periodic full Hugging Face checkpoints when full_hf_checkpoint_interval is configured.
- Drives the vLLM update lifecycle:
    - pause generation
    - flush cache
    - start update
    - install the ModelExpress version
    - finish update
    - resume generation
- Reports delta density, transferred bytes, staging time, publication time, and activation time.

## Configuration
```
  --update-weight-transport modelexpress
  --modelexpress-config '{
    "model_name": "policy",
    "server_url": "modelexpress:8101",
    "initial_base_version_id": "policy-v0",
    "seed_checkpoint_path": "/models/policy",
    "refit_checkpoint_dir": "/tmp/mx-refit",
    "s3_uri_prefix": "s3://weight-updates/policy",
    "s3_endpoint_url": "http://minio:9000",
    "s3_region_name": "us-east-1",
    "full_hf_checkpoint_interval": 10
  }'
```

## Scope

This initial integration supports S3-compatible object storage and the canonical XOR-delta path.
ModelExpress P2P transfer and Dynamo-specific external rollout integration are outside this PR.

## Testing
- 114 targeted unit tests passed.
- Added coverage for:
  - argument parsing and validation
  - lazy updater construction
  - vLLM backend configuration
  - automatic and retryable v0 registration
  - distributed tensor-chunk ownership
  - delta and periodic full-checkpoint publication
  - update lifecycle ordering and failure handling
  - metrics aggregation
- Ruff and git diff --check pass.

## E2E Validation
Ran Qwen3-30B-A3B on B200 nodes with following setup:
- 4 Trainer GPUs (TP2 / PP2 / EP2)
- 4 Rollout GPUs (TP4)
- Run 50 rollout steps
- Full HF checkpoint interval for N=25

Delta
• Versions: v1–v24, v26–v49
• Mean: 46.68s
• Median: 44.94s

Full HF checkpoint
• Versions: v25
• Mean: 545.56s
• Median: 545.56s

### CODE REVIEW

File: vime/backends/megatron_utils/update_weight/update_weight_from_modelexpress.py [hunk 1]
```diff
@@ -0,0 +1,336 @@
...
+
+    def __init__(
+        self,
+        args: Namespace,
+        model: Sequence[torch.nn.Module],
+        weights_getter: Callable[[], Mapping[str, torch.Tensor]],
+        model_name: str,
+        quantization_config: dict[str, int | str | list[str]] | None,
+    ) -> None:
+        self._config = dict(args.modelexpress_config)
```
[the remark this answers is not in the document]
Comment (nv-hwoo) [reply vime/backends/megatron_utils/update_weight/update_weight_from_modelexpress.py:46]: Thanks. modelexpress_config is created and type-checked by VIME before this updater, while ModelExpress owns required-field validation. model_name and server_url also support MX environment/default fallbacks, so I am keeping that validation in MX rather than duplicating it here.

File: vime/backends/megatron_utils/update_weight/update_weight_from_modelexpress.py [hunk 2]
```diff
@@ -0,0 +1,336 @@
...
+        self._object_storage_config = ObjectStorageConfig(
+            storage_type=ObjectStorageType.S3,
+            uri_prefix=self._config.get("s3_uri_prefix", ""),
+            initial_base_version_id=self._config.get(
+                "initial_base_version_id", ""
+            ),
+            seed_checkpoint_path=self._config.get("seed_checkpoint_path", ""),
+            endpoint_url=self._config.get("s3_endpoint_url"),
+            region_name=self._config.get("s3_region_name"),
+        )
```
[the remark this answers is not in the document]
Comment (nv-hwoo) [reply vime/backends/megatron_utils/update_weight/update_weight_from_modelexpress.py:79]: Good catch. Addressed by normalizing explicit null values for s3_uri_prefix, initial_base_version_id, and seed_checkpoint_path to empty strings, with unit coverage.

File: vime/backends/megatron_utils/update_weight/update_weight_from_modelexpress.py [hunk 3]
```diff
@@ -0,0 +1,336 @@
...
+            should_convert_chunk=lambda chunk_idx: chunk_idx % world_size == rank,
+        )
+
+    def _gather_metrics(
+        self,
+        *,
+        group: Any,
+    ) -> dict[str, int | float]:
+        """Aggregate ModelExpress publication and serving-cutover metrics."""
+        local_metrics = self._trainer.pop_metrics()
```
[the remark this answers is not in the document]
Comment (nv-hwoo) [reply vime/backends/megatron_utils/update_weight/update_weight_from_modelexpress.py:272]: ModelExpressTrainerClient.pop_metrics() guarantees a dictionary and returns an empty dictionary when no metrics are available. I am keeping this strict so a None return is surfaced as an MX contract violation rather than masked.

**Comment by aoshen02** [comment]
Hi, It does not need to change the vllm code?

**Comment by nv-hwoo** [comment]
Hi @aoshen02, we use vLLM's [Weight Transfer Engine](https://docs.vllm.ai/en/latest/training/weight_transfer) plugin to define our [custom extension](https://github.com/ai-dynamo/modelexpress/blob/main/modelexpress_client/python/modelexpress_rl/inference/engines/vllm/weight_transfer_engine.py) in our ModelExpress codebase, so we don't need any direct changes in vLLM.

Also, when used with existing ModelExpress client in vLLM engine, we also support P2P RDMA fan-out across the rollout engines during weight update or cold-start (due to crash or restart) as well as S3 fallback. We don't have a clean doc yet but here are the recent PRs:
- https://github.com/ai-dynamo/modelexpress/pull/740
- https://github.com/ai-dynamo/modelexpress/pull/738
- https://github.com/ai-dynamo/modelexpress/pull/727
- https://github.com/ai-dynamo/modelexpress/pull/710

**Comment by aoshen02** [comment]
> Hi @aoshen02, we use vLLM's [Weight Transfer Engine](https://docs.vllm.ai/en/latest/training/weight_transfer) plugin to define our [custom extension](https://github.com/ai-dynamo/modelexpress/blob/main/modelexpress_client/python/modelexpress_rl/inference/engines/vllm/weight_transfer_engine.py) in our ModelExpress codebase, so we don't need any direct changes in vLLM.
> 
> Also, when used with existing ModelExpress client in vLLM engine, we also support P2P RDMA fan-out across the rollout engines during weight update or cold-start (due to crash or restart) as well as S3 fallback. We don't have a clean doc yet but here are the recent PRs:
> 
> * [feat(refit): add P2P-first active refit ai-dynamo/modelexpress#740](https://github.com/ai-dynamo/modelexpress/pull/740)
> * [fix(refit): reconcile distributed RL cold starts ai-dynamo/modelexpress#738](https://github.com/ai-dynamo/modelexpress/pull/738)
> * [feat(refit): add RL cold-start load strategy chain ai-dynamo/modelexpress#727](https://github.com/ai-dynamo/modelexpress/pull/727)
> * [feat(refit): add target-version delta replay ai-dynamo/modelexpress#710](https://github.com/ai-dynamo/modelexpress/pull/710)

Please join this vllm slack channel:
image
